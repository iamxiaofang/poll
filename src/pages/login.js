import { useState } from "react";
import { useDispatch, useSelector } from "react-redux";

export const LoginPage = () => {
  const [username, setUsename] = useState('');
  const [password, setPassword] = useState('');
  const dispatch = useDispatch();

  const handleUsernameChange = (e) => {
    setUsename(e.target.value)
  }
  const handlePasswordChange = (e) => {
    setPassword(e.target.value)
  }
  const handleLogin = (e) => {
    e.preventDefault()

    const user = Object.values(users).find(u => u.id === username && u.password === password)
    if (user) {
      dispatch({
        type: 'LOGIN',
        data: user
      })
    } else {
      alert('Bad login')
    }
  }

  const users = useSelector((state) => state.users)

  return (
    <form>
      <p>
        <label htmlFor="username">Username: </label>
        <input data-testid="username" onChange={handleUsernameChange} value={username} name="username" text="text" />
      </p>

      <p>
        <label htmlFor="password">Password: </label>
        <input data-testid="password" onChange={handlePasswordChange} value={password} name="password" type="text" />
      </p>

      <button data-testid="login-button" onClick={handleLogin}>Login</button>
    </form>
  )
}