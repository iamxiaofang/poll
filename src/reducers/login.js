export default function login(state = {
  user: null
}, action) {
  switch (action.type) {
    case 'LOGIN':
      return {
        ...state,
        user: action.data
      }
    case 'LOGOUT':
      return {
        ...state,
        user: null,
      }
    default:
      return state;
  }
}
