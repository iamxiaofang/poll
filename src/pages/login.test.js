import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router';
import configureMockStore from 'redux-mock-store';
import { thunk } from "redux-thunk";
import { LoginPage } from './login';


const mockStore = configureMockStore([thunk]);

describe('LoginPage', () => {

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('matches snapshot', () => {

    const store = mockStore({
      users: {
        'xiaofang': {
          id: 'xiaofang',
          password: 'password123',
          name: '小芳',
          avatarURL: 'https://example.com/avatar.jpg',
          answers: {},
          questions: []
        }
      }
    });

    const page = render(
      <Provider store={store} >
        <MemoryRouter>
          <LoginPage />
        </MemoryRouter>
      </Provider>
    );
    expect(page).toMatchSnapshot();
  });

  test('Login success with good username and password', () => {

    const user = {
      id: 'xiaofang',
      password: 'password123',
      name: '小芳',
      avatarURL: 'https://example.com/avatar.jpg',
      answers: {},
      questions: []
    }

    const store = mockStore({
      users: { [user.id]: user }
    });

    const page = render(
      <Provider store={store} >
        <MemoryRouter>
          <LoginPage />
        </MemoryRouter>
      </Provider>
    );

    const usernameInput = page.getByTestId('username')
    const passwordInput = page.getByTestId('password')
    const loginButton = page.getByTestId('login-button')

    const dispatch = jest.spyOn(store, 'dispatch');

    fireEvent.change(usernameInput, { target: { value: 'xiaofang' } })
    fireEvent.change(passwordInput, { target: { value: 'password123' } })
    fireEvent.click(loginButton)

    expect(dispatch).toHaveBeenCalledWith({
      type: 'LOGIN',
      data: user
    });
  });

  test('Login fails with bad username and password', () => {

    const mockAlert = jest.fn()
    jest.spyOn(window, 'alert').mockImplementation(mockAlert);

    const user = {
      id: 'xiaofang',
      password: 'password123',
      name: '小芳',
      avatarURL: 'https://example.com/avatar.jpg',
      answers: {},
      questions: []
    }

    const store = mockStore({
      users: { [user.id]: user }
    });

    const page = render(
      <Provider store={store} >
        <MemoryRouter>
          <LoginPage />
        </MemoryRouter>
      </Provider>
    );

    const usernameInput = page.getByTestId('username')
    const passwordInput = page.getByTestId('password')
    const loginButton = page.getByTestId('login-button')

    fireEvent.change(usernameInput, { target: { value: 'xiaofang' } })
    fireEvent.change(passwordInput, { target: { value: 'wrong' } })
    fireEvent.click(loginButton)

    expect(mockAlert).toHaveBeenCalledWith('Bad login');
  });

});