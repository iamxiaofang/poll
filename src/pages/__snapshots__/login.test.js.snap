// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form>
        <p>
          <label
            for="username"
          >
            Username: 
          </label>
          <input
            data-testid="username"
            name="username"
            text="text"
            value=""
          />
        </p>
        <p>
          <label
            for="password"
          >
            Password: 
          </label>
          <input
            data-testid="password"
            name="password"
            type="text"
            value=""
          />
        </p>
        <button
          data-testid="login-button"
        >
          Login
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <form>
      <p>
        <label
          for="username"
        >
          Username: 
        </label>
        <input
          data-testid="username"
          name="username"
          text="text"
          value=""
        />
      </p>
      <p>
        <label
          for="password"
        >
          Password: 
        </label>
        <input
          data-testid="password"
          name="password"
          type="text"
          value=""
        />
      </p>
      <button
        data-testid="login-button"
      >
        Login
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
